Runtime error using forward referrence

&gt; cat t5.java
```java
public class t5 {
  static String f = myForwardRef();

  static String myForwardRef(){
    return ("Nigel is best");
  }
  public static void main (String[] args){
    System.out.println(f);
  }
}
```
Process cat exited with code 0

Java produces the familiar:
&gt; javac t5.java
Process javac exited with code 0
&gt; java t5
Nigel is best
Process java exited with code 0

BeanShell will do anything to avoid saying it:
&gt; java -classpath ../../bsh-2.0b4.jar bsh.Interpreter 
t5.java
Script threw exception: Sourced file: t5.java : Typed 
variable declaration : Exception invoking imported object 
method. : at Line: 2 : in file: t5.java : myForwardRef \( \) 

Target exception: 
java.lang.reflect.InvocationTargetException

java.lang.reflect.InvocationTargetException
at 
sun.reflect.NativeMethodAccessorImpl.invoke0\(Native 
Method\)
at 
sun.reflect.NativeMethodAccessorImpl.invoke
\(NativeMethodAccessorImpl.java:39\)
at 
sun.reflect.DelegatingMethodAccessorImpl.invoke
\(DelegatingMethodAccessorImpl.java:25\)
at java.lang.reflect.Method.invoke
\(Method.java:585\)
at bsh.Reflect.invokeMethod\(Unknown 
Source\)
at bsh.BshMethod.invoke\(Unknown Source\)
at bsh.BshMethod.invoke\(Unknown Source\)
at bsh.Name.invokeLocalMethod\(Unknown 
Source\)
at bsh.Name.invokeMethod\(Unknown Source\)
at bsh.BSHMethodInvocation.eval\(Unknown 
Source\)
at bsh.BSHPrimaryExpression.eval\(Unknown 
Source\)
at bsh.BSHPrimaryExpression.eval\(Unknown 
Source\)
at bsh.BSHVariableDeclarator.eval\(Unknown 
Source\)
at bsh.BSHTypedVariableDeclaration.eval
\(Unknown Source\)
at bsh.BSHBlock.evalBlock\(Unknown 
Source\)
at 
bsh.ClassGeneratorImpl.generateClassImpl\(Unknown 
Source\)
at bsh.ClassGeneratorImpl.generateClass
\(Unknown Source\)
at bsh.BSHClassDeclaration.eval\(Unknown 
Source\)
at bsh.Interpreter.eval\(Unknown Source\)
at bsh.Interpreter.source\(Unknown Source\)
at bsh.Interpreter.main\(Unknown Source\)
Caused by: java.lang.NullPointerException
at t5.myForwardRef\(BeanShell Generated via 
ASM \(www.objectweb.org\)\)
... 21 more
Process java exited with code 0


Reported by: nigel_galloway

Original Ticket: "beanshell/bugs/221":https://sourceforge.net/p/beanshell/bugs/221

Exception in BeanShell 3.0.0-SNAPSHOT.841
[t5.java.txt](https://github.com/beanshell/beanshell/files/10145526/t5.java.txt)


```java
bsh % source("t5.java");
--> $0 = class t5 :Class
bsh % t5.main(new String[0]);
// Error: Internal Error: Exception in static init block <clinit> for class t5. With message: Typed variable declaration : Exception invoking imported object method. : at Line: 2 : in file: /home/intriazen/git/beanshell/t5.java : myForwardRef ( )
Caused by: java.lang.reflect.InvocationTargetException
bsh.EvalError: Method myForwardRef() not found in bsh scripted object: t5 : at Line: -1 : in file: <Called from Java Code> : <Compiled Java Code>
: <at unknown location>

bsh.EvalError: Method myForwardRef() not found in bsh scripted object: t5 : at Line: -1 : in file: <Called from Java Code> : <Compiled Java Code>
```

This is not just related to static members but instance references will also fail.

```java
BeanShell 3.0.0-SNAPSHOT.3703
class A { int g = get(); int get() { return 1; } }
--> $0 = class A :Class
new A().g;
// Error: Target Exception: Object constructor : at Line: 2 : in file: <unknown file> : new A ( ) 
Caused by: bsh.InterpreterError: Error in class instance initialization: bsh.InterpreterError: Error in class instance This initialization: bsh.TargetError: Typed variable declaration : Exception invoking imported object method. : at Line: 1 : in file: <unknown file> : get ( ) 
Caused by: java.lang.reflect.InvocationTargetException
bsh.EvalError: Method get() not found in bsh scripted object: A : at Line: -1 : in file: <Called from Java Code> : <Compiled Java Code>
```

--- a/src/conf/version.properties
+++ b/src/conf/version.properties
@@ -14,6 +14,6 @@
 #
 #Project version number - populated by maven
 #Project build number - incremented by beanshell-maven-plugin
-#Thu Jan 12 08:45:42 SAST 2023
-build=3703
+#Thu Jan 12 11:23:37 SAST 2023
+build=3761
 release=${project.version}

--- a/src/main/java/bsh/ClassGenerator.java
+++ b/src/main/java/bsh/ClassGenerator.java
@@ -220,16 +220,24 @@ public final class ClassGenerator {
      * members are passed, etc.
      */
     static class ClassNodeFilter implements BSHBlock.NodeFilter {
-        enum Context { STATIC, INSTANCE, CLASSES }
-
-        public static ClassNodeFilter CLASSSTATIC = new ClassNodeFilter(Context.STATIC);
-        public static ClassNodeFilter CLASSINSTANCE = new ClassNodeFilter(Context.INSTANCE);
+        private enum Context { STATIC, INSTANCE, CLASSES }
+        private enum Types { ALL, METHODS, FIELDS }
+        public static ClassNodeFilter CLASSSTATICFIELDS = new ClassNodeFilter(Context.STATIC, Types.FIELDS);
+        public static ClassNodeFilter CLASSSTATICMETHODS = new ClassNodeFilter(Context.STATIC, Types.METHODS);
+        public static ClassNodeFilter CLASSINSTANCEFIELDS = new ClassNodeFilter(Context.INSTANCE, Types.FIELDS);
+        public static ClassNodeFilter CLASSINSTANCEMETHODS = new ClassNodeFilter(Context.INSTANCE, Types.METHODS);
         public static ClassNodeFilter CLASSCLASSES = new ClassNodeFilter(Context.CLASSES);
 
         Context context;
+        Types types = Types.ALL;
 
         private ClassNodeFilter(Context context) {
             this.context = context;
+        }
+
+        private ClassNodeFilter(Context context, Types types) {
+            this.context = context;
+            this.types = types;
         }
 
         @Override
@@ -239,31 +247,42 @@ public final class ClassGenerator {
             // Only show class decs in CLASSES
             if (node instanceof BSHClassDeclaration) return false;
 
-            if (context == Context.STATIC) return isStatic(node);
+            if (context == Context.STATIC)
+                return types == Types.METHODS ? isStaticMethod(node) : isStatic(node);
 
-            if (context == Context.INSTANCE) return !isStatic(node);
-
-            // ALL
-            return true;
+            // context == Context.INSTANCE cannot be anything else
+            return types == Types.METHODS ? isInstanceMethod(node) : isNonStatic(node);
         }
 
-        boolean isStatic(Node node) {
-            if ( null != node.jjtGetParent()
-                    && node.jjtGetParent().jjtGetParent() instanceof BSHClassDeclaration
+        private boolean isStatic(Node node) {
+            if ( node.jjtGetParent().jjtGetParent() instanceof BSHClassDeclaration
                     && ((BSHClassDeclaration) node.jjtGetParent().jjtGetParent()).type == Type.INTERFACE )
                 return true;
 
             if (node instanceof BSHTypedVariableDeclaration)
-                return ((BSHTypedVariableDeclaration) node).modifiers != null
-                    && ((BSHTypedVariableDeclaration) node).modifiers.hasModifier("static");
-
-            if (node instanceof BSHMethodDeclaration)
-                return ((BSHMethodDeclaration) node).modifiers != null
-                    && ((BSHMethodDeclaration) node).modifiers.hasModifier("static");
+                return ((BSHTypedVariableDeclaration) node).modifiers.hasModifier("static");
 
             if (node instanceof BSHBlock)
                 return ((BSHBlock) node).isStatic;
 
+            return false;
+        }
+
+        private boolean isNonStatic(Node node) {
+            if (node instanceof BSHMethodDeclaration)
+                return false;
+            return !isStatic(node);
+        }
+
+        private boolean isStaticMethod(Node node) {
+            if (node instanceof BSHMethodDeclaration)
+                return ((BSHMethodDeclaration) node).modifiers.hasModifier("static");
+            return false;
+        }
+
+        private boolean isInstanceMethod(Node node) {
+            if (node instanceof BSHMethodDeclaration)
+                return !((BSHMethodDeclaration) node).modifiers.hasModifier("static");
             return false;
         }
     }

--- a/src/main/java/bsh/This.java
+++ b/src/main/java/bsh/This.java
@@ -28,8 +28,10 @@
 
 package bsh;
 
-import static bsh.ClassGenerator.ClassNodeFilter.CLASSINSTANCE;
-import static bsh.ClassGenerator.ClassNodeFilter.CLASSSTATIC;
+import static bsh.ClassGenerator.ClassNodeFilter.CLASSINSTANCEFIELDS;
+import static bsh.ClassGenerator.ClassNodeFilter.CLASSINSTANCEMETHODS;
+import static bsh.ClassGenerator.ClassNodeFilter.CLASSSTATICFIELDS;
+import static bsh.ClassGenerator.ClassNodeFilter.CLASSSTATICMETHODS;
 import static bsh.ClassGeneratorUtil.DEFAULTCONSTRUCTOR;
 import static bsh.This.Keys.BSHCONSTRUCTORS;
 import static bsh.This.Keys.BSHINIT;
@@ -760,8 +762,9 @@ public final class This implements java.io.Serializable, Runnable
             }
 
             // evaluate the instance portion of the block in it
-            try { // Evaluate the initializer block
-                instanceInitBlock.evalBlock(new CallStack(instanceNameSpace), instanceThis.declaringInterpreter, true/*override*/, CLASSINSTANCE);
+            try { // Evaluate the initializer block methods first
+                instanceInitBlock.evalBlock(new CallStack(instanceNameSpace), instanceThis.declaringInterpreter, true/*override*/, CLASSINSTANCEMETHODS);
+                instanceInitBlock.evalBlock(new CallStack(instanceNameSpace), instanceThis.declaringInterpreter, true/*override*/, CLASSINSTANCEFIELDS);
             } catch (Exception e) {
                 throw new InterpreterError("Error in class instance This initialization: " + e, e);
             }
@@ -790,8 +793,9 @@ public final class This implements java.io.Serializable, Runnable
             BSHBlock block = (BSHBlock) classStaticNameSpace.getVariable(BSHINIT.toString());
             CallStack callstack = new CallStack(classStaticNameSpace);
 
-            // evaluate the static portion of the block in the static space
-            block.evalBlock(callstack, interpreter, true/*override*/, CLASSSTATIC);
+            // evaluate the static portion of the block in the static space methods first
+            block.evalBlock(callstack, interpreter, true/*override*/, CLASSSTATICMETHODS);
+            block.evalBlock(callstack, interpreter, true/*override*/, CLASSSTATICFIELDS);
 
             // Validate that static final variables were set
             for (Variable var : Reflect.getVariables(classStaticNameSpace))

--- a/src/test/resources/test-scripts/class17.bsh
+++ b/src/test/resources/test-scripts/class17.bsh
@@ -1,0 +1,50 @@
+source("TestHarness.bsh");
+source("Assert.bsh");
+
+/*
+    This test case looks at method references declared after referenced in fields. #359
+*/
+
+class MyStatic {
+    static int g = get();
+    static String f;
+    static {
+        f = fetch();
+    }
+    static int get() {
+        return 22;
+    }
+    static String fetch() {
+        return "22";
+    }
+}
+
+assertEquals("Integer g is 22", 22, MyStatic.g);
+assertNotNull("String f is set through static block and should not be null", MyStatic.f);
+assertEquals("String f is '22'", "22", MyStatic.f);
+assertEquals("String fetch is '22'", "22", MyStatic.fetch());
+assertEquals("Integer get is 22", 22, MyStatic.get());
+
+class MyInstance {
+    int g = get();
+    String f;
+    MyInstance() {
+        f = fetch();
+    }
+    int get() {
+        return 22;
+    }
+    String fetch() {
+        return "22";
+    }
+}
+
+myInstance = new MyInstance();
+assertEquals("Integer g is 22", 22, myInstance.g);
+assertNotNull("String f is set through constructor and should not be null", myInstance.f);
+assertEquals("String f is '22'", "22", myInstance.f);
+assertEquals("String fetch is '22'", "22", myInstance.fetch());
+assertEquals("Integer get is 22", 22, myInstance.get());
+
+complete();
+

